File: src/main/java/net/hectus/neobb/game/BossBarGame.java
package net.hectus.neobb.game;

import net.hectus.neobb.turn.default_game.warp.WarpTurn;
import net.kyori.adventure.bossbar.BossBar;
import net.kyori.adventure.text.Component;
import org.bukkit.World;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public abstract class BossBarGame extends Game {
    protected BossBar bossBar;

    public BossBarGame(boolean ranked, World world, @NotNull List<Player> players, WarpTurn defaultWarp) {
        super(ranked, world, players, defaultWarp);
    }

    public BossBar initialBossBar() {
        return BossBar.bossBar(Component.text("Turn Countdown: -"), 0.0f, BossBar.Color.YELLOW, BossBar.Overlay.NOTCHED_10);
    }

    public void bossBar(@NotNull BossBar bossBar) {
        bossBar.name(Component.text("Turn Countdown: " + turnCountdown + "s"));
        bossBar.progress(Math.clamp((float) turnCountdown / info().turnTimer(), 0.0f, 1.0f));
    }

    public final BossBar bossBar() {
        return bossBar;
    }

    @Override
    public void start() {
        super.start();
        bossBar = initialBossBar();
        bossBar.addViewer(gameTarget(false));
    }

    @Override
    public void end(boolean force) {
        bossBar.removeViewer(gameTarget(false));
        super.end(force);
    }
}

File: src/main/kotlin/net/hectus/neobb/game/BossBarGame.kt
package net.hectus.neobb.game

import com.marcpg.libpg.display.SimpleBossBar
import com.marcpg.libpg.util.component
import net.hectus.neobb.game.util.GameDifficulty
import net.kyori.adventure.bossbar.BossBar
import org.bukkit.World
import org.bukkit.entity.Player

abstract class BossBarGame(world: World, bukkitPlayers: List<Player>, difficulty: GameDifficulty = GameDifficulty.NORMAL) : Game(world, bukkitPlayers, difficulty) {
    open val bossBar: SimpleBossBar = SimpleBossBar(target(false), 20, { component("Turn Countdown: ${turnCountdown}s") }, { (turnCountdown.toFloat() / (info.turnTimer * difficulty.timeFactor).toInt()).coerceIn(0.0f, 1.0f) }, { BossBar.Color.YELLOW }, { BossBar.Overlay.NOTCHED_10 })

    override fun extraStart() = bossBar.start()
    override fun extraEnd(force: Boolean) = bossBar.stop()
}
